src/module_4/module_4_model-fitting.md

```diff
@@ -0,0 +1,1214 @@
+
+    ax[0, 0].plot(recall, precision, label=f"{model_name} AUC: {pr_auc:.2f}")
+    ax[0, 0].set_xlabel("Recall")
+    ax[0, 0].set_ylabel("Precision")
+    ax[0, 0].set_title(f"Precision-Recall Curve")
+    ax[0, 0].legend()
+
+    ax[0, 1].plot(fpr, tpr, label=f"AUC: {roc_auc:.2f}")
+    ax[0, 1].set_xlabel("FPR")
+    ax[0, 1].set_ylabel("TPR")
+    ax[0, 1].set_title(f"ROC Curve")
+    ax[0, 1].legend()
+
+    ax[1, 0].axis('off')  # No se necesita eje para mostrar texto
+    ax[1, 0].text(0.5, 0.5, f"Accuracy: {accuracy:.2f}", fontsize=15, ha='center')
+
+    sns.heatmap(conf_matrix, annot=True, fmt='d', cmap='Blues', ax=ax[1, 1])
+    ax[1, 1].set_title("Confusion Matrix")
+    ax[1, 1].set_xlabel("Predicted label")
+    ax[1, 1].set_ylabel("True label")
+
+    plt.tight_layout()
+    plt.show()
+    print("Precision Recall AUC: " + str(pr_auc))
+    print("Roc Curve AUC: " + str(roc_auc))
+    print("Sum of both AUCs: " + str(pr_auc + roc_auc))
+```
+
+
+```python
+"""""Función para ver la importancia de los coeficientes"""
+def plot_feature_importance(clf, columns):
+    fig, ax = plt.subplots()
+    fi = pd.DataFrame(list(zip(columns, clf.feature_importances_)), columns=["features", "importance"])
+    fi.sort_values(by="importance", ascending=True, inplace=True)
+    fi.plot(kind="barh", x="features", y="importance", ax=ax)
+    return fig, ax
+```
+
+## Previous Baselines
+
+### Initial Benchmark 
+
+Para nuestra evaluación inicial, utilizamos la característica de "popularidad global" para determinar si el pedido puede ser realizado o no.
+
+
+```python
+plot_metrics("Baseline with global popularity",val_df["global_popularity"],val_df[label_col])
+```
+
+
+    
+![png](module_4_model-fitting_files/module_4_model-fitting_10_0.png)
+    
+
+
+### Best Linear models result
+
+Para nuestra segunda evaluación, emplearemos el modelo lineal que mostró mejores resultados en el ejercicio anterior: regresión Lasso sin variables categóricas, eliminando variables con una importancia menor a 0.02.
+
+
+```python
+train_cols = ["days_since_purchase_variant_id", "discount_pct", "std_days_to_buy_product_type", "count_pets", "set_as_regular", "avg_days_to_buy_product_type", "days_since_purchase_product_type", "user_order_seq", "avg_days_to_buy_variant_id", "abandoned_before", "global_popularity", "normalised_price", "ordered_before",]
+
+regularization_values = [1e-4, 1e-3, 1e-2, 1e-1, 1, 10]
+
+best_auc = 0
+best_c = None
+best_model = None
+
+for c in regularization_values:
+    lr = make_pipeline(
+        StandardScaler(),
+        LogisticRegression(penalty='l1',C=c, solver='liblinear')
+    )
+    lr.fit(X_train[train_cols], y_train)
+
+    y_proba = lr.predict_proba(X_test[train_cols])[:, 1]
+    
+    roc_auc = roc_auc_score(y_test, y_proba)
+    
+    if roc_auc > best_auc:
+        best_auc = roc_auc
+        best_c = c
+        best_model = lr
+
+y_pred = best_model.predict_proba(X_test[train_cols])[:, 1]
```

Comment:
instead of visualizing just the best_model metrics why don't you look at all of them? this way you can understand if there is a big difference / variance between the model results (and therefore finnetuning even further makes sense) or if doesn't really matter much